Katello/katello
# Fixes #27497 - Support dmi.system.uuid override

In some circumstances content hosts can report the same `dmi.system.uuid` fact as other hosts. This fact needs to be unique and this PR provides a way to work around the problem as a stopgap until the problem is (ideally) fixed in the infrastructure.

Also in this PR is a relocation and minor refactor of the registration bits in `SubscriptionFacet`. I've moved most of the code to `RegistrationManager` and split up the methods slightly to reflect the logical parts of the registration which makes thinking about, reading, and testing the code much easier. Also removed some dead code.

**How does it work?**
A new setting is provided which allows users to input the DMI UUID which is being reported by multiple systems. Doing so will have the following effects:
- During registration a random UUID will be generated and its value sent to Candlepin. It will also be stored as a host parameter for reference during checkin
- During checkin, if the host is reporting one of the duplicate values, the value is taken from the host parameter instead of what was already in the facts. If no host parameter exists (accidental deletion, whatever) a new one is generated and stored as a host parameter.

**To test**
- Register a content host (ex katello-client from forklift)
- Run `subscription-manager clean` on the client so it believes it's unregistered but preserves its data on the server
- On the same client, change its hostname so that registering will attempt to create a new profile.
- Attempt to register (`subscription-manager register`) it should fail with an error

Now, we can work around this problem.

- Take the dmi.system.uuid from the client: `subscription-manager facts | grep dmi.system.uuid`
- In the UI, go to Administer -> Settings -> Content
- Find the new setting 'Host Duplicate DMI UUIDs' and input the UUID you found in array format
- Registering the system should work, and it should have a new DMI UUID on the server - not on the client

You can verify that with some queries:

```
::Host.find(197).fact_values.where(fact_name_id: Katello::RhsmFactName.find_by_name('dmi::system::uuid'))

::Host.find(197).subscription_facet.candlepin_consumer.consumer_attributes[:facts]['dmi.system.uuid']

::Host.find(197).parameters

```

The UUIDs should all match.

- From the client, refresh the facts `subscription-manager facts --update`
- The UUID on the server side should be unchanged
- Deleting the host parameter and refreshing facts again should generate and store and new UUID
